### PR TITLE
fix  main thread was stuck

### PR DESCRIPTION
**Pull request by wangyuanyuanyy**
Fix main thread was stuck caused by `[NSData dataWithContentsOfURL:url]`

### CODE REVIEW

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 1]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
 
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
-    }
-  }
-  return _icon;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:108]: Why is this removed?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 2]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
+- (void)loadRequireData {
+    NSString *urlString = _nativeAd.data.icon.imageURL;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:97]: Is it guaranteed that we can get the `imageURL` successfully every time through accessing the property?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 3]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
+    dispatch_async(dispatch_get_global_queue(0, 0), ^{
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:98]: Seems like we are having the priority `DISPATCH_QUEUE_PRIORITY_DEFAULT` here? Can you explain why this is chosen against the other priorities (e.g. `DISPATCH_QUEUE_PRIORITY_HIGH`)?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 4]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
-    }
-  }
-  return _icon;
+- (void)loadRequireData {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: We probably can move all the implementations to above (i.e. line `92`) without having an extra method.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 5]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
+    NSString *urlString = _nativeAd.data.icon.imageURL;
+    dispatch_async(dispatch_get_global_queue(0, 0), ^{
+        NSURL *url = [NSURL URLWithString:urlString];
+        NSData *data = [NSData dataWithContentsOfURL:url];
+        GADPangleRTBNativeRenderer *__weak weakSelf = self;
+        dispatch_async(dispatch_get_main_queue(), ^{
+            GADPangleRTBNativeRenderer *strongSelf = weakSelf;
+            if (!strongSelf) {
+               return;
+            }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:106]: Do we really need the `weakSelf` and `strongSelf` implementations here? We already do that above.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 2 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:97]: Do we need the check below?
```
if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil)
```

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 6]
```diff
@@ -180,13 +189,6 @@ - (BOOL)handlesUserImpressions {
   return YES;
 }
 
-- (GADNativeAdImage *)imageWithUrlString:(NSString *)urlString {
-  NSURL *url = [NSURL URLWithString:urlString];
-  NSData *data = [NSData dataWithContentsOfURL:url];
-  UIImage *image = [UIImage imageWithData:data];
-  return [[GADNativeAdImage alloc] initWithImage:image];
-}
-
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: Why is this method removed? Seems like we can use it above?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 7]
```diff
@@ -88,26 +88,35 @@ - (void)renderNativeAdForAdConfiguration:
...
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
-    }
-  }
-  return _icon;
+- (void)loadRequireData {
+    NSString *urlString = _nativeAd.data.icon.imageURL;
+    dispatch_async(dispatch_get_global_queue(0, 0), ^{
+        NSURL *url = [NSURL URLWithString:urlString];
+        NSData *data = [NSData dataWithContentsOfURL:url];
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:100]: You mentioned that the thread was stuck because of this line. Can you clarify about this? Should there be a fix in Pangle SDK if the url cannot be populated?

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:108]: `Don't use this synchronous method to request network-based URLs. For network-based URLs, this method can block the current thread for tens of seconds on a slow network, resulting in a poor user experience, and in iOS, may cause your app to be terminated.` in apple development doc

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 6 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: It moved in `- (void)loadRequireData`

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 5 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: I will remove it.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 4 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: Will this nest too many levels? Code is less readable.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 3 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 2 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:97]: I changed the code.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 1 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: The `_icon` set in `-loadRequireData`

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 8]
```diff
@@ -88,26 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
-                   strongSelf->_delegate = delegate;
-                 }
+      
+                 [strongSelf loadRequireData];
                }];
 }
 
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:102]: We need this method to map the `icon` asset.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 9]
```diff
@@ -88,26 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+- (void)loadRequireData {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: I think we should still move this method to above.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 10]
```diff
@@ -88,26 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
+        return;
     }
-  }
-  return _icon;
+    NSURL *url = [NSURL URLWithString:urlString];
+    if (!url) {
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
+        return;
+    }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:106]: We shouldn't need this check if `urlString` is valid?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 11]
```diff
@@ -88,26 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:107]: Can you explain why it is `HIGH` now?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 7 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:108]: Seems like this is the documentation: https://developer.apple.com/documentation/foundation/nsdata/1547245-datawithcontentsofurl

Then shouldn't we use `[dataTaskWithURL:completionHandler:](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl)` instead?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 6 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: We can keep this method.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 12]
```diff
@@ -88,26 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+- (void)loadRequireData {
+    NSString *urlString = _nativeAd.data.icon.imageURL;
+    if (!urlString.length) {
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
+        return;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:100]: What to do with the `icon` in this case?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 8 above)
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:102]: Also, shouldn't all the changes / fixes be added here instead since the issue is related to the icon?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 6 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: This method  was used to load icon image, but now the loading progress is implemented like this:
> GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
> self->_icon = image;

so we don't need this method anymore

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 2 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:97]: `- loadRequireData` is invoked when PAGLNativeAd loadAd callback success.So _nativeAd.data isn't nil.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 7 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:108]: We use `-dataWithContentsOfURL` asynchronously. Must we use  `[dataTaskWithURL:completionHandler:]`?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 8 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:102]: `_icon` set is implement like this, in method`- (void)loadRequireData`
> GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
>  self->_icon = image;

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 9 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: You mean moving  `-loadRequireData`  line 96 to  line 105?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 10 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:106]: `[NSURL URLWithString:urlString]` is return `An NSURL object initialized with URLString. If the URL string was malformed or nil, returns nil.`

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 12 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:100]: In this case icon is nil. In `GADMediationNativeAd` protocol, icon is a nullable property.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 8 above)
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:102]: As the previous comment was resolved, I am going to add another one.

Please keep this method as it's for mapping the icon asset, also see [the API documentation](https://developers.google.com/ad-manager/mobile-ads-sdk/ios/api/reference/Protocols/GADMediatedUnifiedNativeAd.html#icon).

We should do:
```
- (nullable GADNativeAdImage *)icon {
  return _icon;
}
```

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:102]: ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 13]
```diff
@@ -88,23 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
...
+              self->_delegate = delegate;
+            }
+        });
+    });
+}
+
+#pragma mark - GADMediationNativeAd
+- (nullable GADNativeAdImage *)icon {
   return _icon;
 }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:128]: For all the assets, can you add a `nullable` annotation just like what you did for `icon`?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 14]
```diff
@@ -180,13 +198,6 @@ - (BOOL)handlesUserImpressions {
   return YES;
 }
 
-- (GADNativeAdImage *)imageWithUrlString:(NSString *)urlString {
-  NSURL *url = [NSURL URLWithString:urlString];
-  NSData *data = [NSData dataWithContentsOfURL:url];
-  UIImage *image = [UIImage imageWithData:data];
-  return [[GADNativeAdImage alloc] initWithImage:image];
-}
-
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: Can we do the below instead?
```
- (GADNativeAdImage *)loadIconWithUrlString:(NSString *)urlString {
  NSURL *url = [NSURL URLWithString:urlString];
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
    NSData *data = [NSData dataWithContentsOfURL:url];
    dispatch_async(dispatch_get_main_queue(), ^{
      GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
    });
    _icon = image;
  });
}

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 15]
```diff
@@ -88,23 +88,41 @@ - (void)renderNativeAdForAdConfiguration:
                  strongSelf->_nativeAd = nativeAd;
                  strongSelf->_nativeAd.delegate = strongSelf;
                  strongSelf->_nativeAd.rootViewController = adConfiguration.topViewController;
-
-                 if (strongSelf->_loadCompletionHandler) {
-                   id<GADMediationNativeAdEventDelegate> delegate =
-                       strongSelf->_loadCompletionHandler(strongSelf, nil);
-                   strongSelf->_delegate = delegate;
-                 }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: I have added a new method `loadIconWithUrlString` in another comment.
Then in here can we revert this change and call that method instead?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 14 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: `_icon = image` will report an error because the image is not within the scope of the current position. And this method will cause a delay in getting the icon.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 15 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: I replied to your comment.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 13 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:128]: ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 14 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:189]: In the snippet, the `_icon = nil` obtained when invoke `_loadCompletionHandler`

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 16]
```diff
@@ -88,79 +88,97 @@ - (void)renderNativeAdForAdConfiguration:
...
+            }
+            GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
+            self->_icon = image;
+            if (self->_loadCompletionHandler) {
+              id<GADMediationNativeAdEventDelegate> delegate =
+                  self->_loadCompletionHandler(self, nil);
+              self->_delegate = delegate;
+            }
+        });
+    });
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:121]: @wangyuanyuanyy 

Since the URL is a network URL, can you update the implementation with [dataTaskWithURL:completionHandler:](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl) method of the [NSURLSession](https://developer.apple.com/library/archive/technotes/tn2232/_index.html#//apple_ref/doc/uid/DTS40012884-CH1-SECNSURLSESSION) class?

Even on concurrent queues, running long synchronous methods is not ideal. Since there's a limit of 64 concurrent processes across all concurrent queues and it can fill up suprisingly quick with long sync calls.

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:121]: Do you mean the code is written like this ?
> NSURLSession *session = [NSURLSession sharedSession];
    [session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (!self) {
                return;
            }
            GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
            self->_icon = image;
            if (self->_loadCompletionHandler) {
              id<GADMediationNativeAdEventDelegate> delegate =
                  self->_loadCompletionHandler(self, nil);
              self->_delegate = delegate;
            }
        });
    }];

Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:121]: @wangyuanyuanyy I haven't looked into the detailed implementation yet, but once you make the change to use [dataTaskWithURL:completionHandler:](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl) method of the [NSURLSession](https://developer.apple.com/library/archive/technotes/tn2232/_index.html#//apple_ref/doc/uid/DTS40012884-CH1-SECNSURLSESSION) class, we can review the code again.

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: Ok~ I changed .

**Comment by wangyuanyuanyy** [comment]
Hello, I changed method of the load image data. Please review the again.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 17]
```diff
@@ -88,79 +88,98 @@ - (void)renderNativeAdForAdConfiguration:
...
 }
 
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+- (void)loadRequireData {
+    NSString *urlString = _nativeAd.data.icon.imageURL;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:97]: `URLString`, acronyms in ObjC are always uppercase.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 18]
```diff
@@ -88,79 +88,98 @@ - (void)renderNativeAdForAdConfiguration:
...
+    if (!urlString.length) {
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:99]: Rather than needing to wrap this in every location, we can create a block locally at the top like:

```
localBlock = ^{(NSError *error)if (_loadCompletionHandler) {
   _loadCompletionHandler(self,error);
}};
```
then everywhere below can call localBlock without a guard and not risk crashing.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 19]
```diff
@@ -88,79 +88,98 @@ - (void)renderNativeAdForAdConfiguration:
...
                }];
 }
 
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+- (void)loadRequireData {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:96]: Nit, `loadRequiredData` or `loadNativeData`.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 17 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 19 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: Ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m (hunk 18 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: Ok~

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 20]
```diff
@@ -88,79 +88,101 @@ - (void)renderNativeAdForAdConfiguration:
...
-  }
+    NSURL *url = [NSURL URLWithString:URLString];
+    if (!url) {
+        localBlock();
+        return;
+    }
+    NSURLSession *session = [NSURLSession sharedSession];
+    NSURLSessionDataTask *task = [session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            if (!self) {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:117]: I assume that this is intended to be another `weakSelf` / `strongSelf` construct. `self` will always be `nonnull` here because this is retaining it.

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m]: ok~
